File: Parser.java
import java.util.ArrayList;

abstract public class Parser {
	protected ArrayList<Token> tokens;
	protected SymbolsTable symbolsTable;
	protected ArrayList<Tuple> tuples = new ArrayList<Tuple>();
	protected int tokenIndex = 0;
	protected int tokenIndexAux;
	protected boolean debug;

	abstract public boolean parse(ArrayList<Token> tokens, SymbolsTable symbolsTable, boolean debug);

	public ArrayList<Token> getTokens() {
		return tokens;
	}

	public SymbolsTable getSymbolsTable() {
		return symbolsTable;
	}

	public ArrayList<Tuple> getTuples() {
		return tuples;
	}

	protected boolean match(String name) {
        if (tokenIndex >= tokens.size())
            return false;
        
		if (tokens.get(tokenIndex).getType().getName().equals(name)) {
			tokenIndex++;
			
			if (debug)
				System.out.println("Match: " + name);

			return true;
		}

		return false;
	}
	
	protected int resetIndexes(int tokenIndex, int tupleIndex) {
	    this.tokenIndex = tokenIndex;
	    deleteTuples(tupleIndex, tuples.size());
	    return tuples.size();
	}

	protected void deleteTuples(int begin, int end) {
		int n = end - begin;
		
		for (int i=0; i<n; i++)
			tuples.remove(tuples.size()-1);
	}

	protected void buildTuple(String type, int beginToken, int endToken, int nextTrue, int nextFalse) {
		ArrayList<Token> elements = new ArrayList<Token>();

		for (int i=beginToken; i<endToken; i++)
			elements.add(tokens.get(i));

		//tuples.add(new Tuple(type, elements, beginToken, nextTrue, nextFalse));
		tuples.add(new Tuple(type, elements, nextTrue, nextFalse));
	}
}

File: Parser.swift
import Foundation

class Parser {
    var tokens: [Token] = []
    var symbols: SymbolsTable = SymbolsTable()
    var tuples: [Tuple] = []
    var tokenIndex = 0
    var debug: Bool = true
    
    init() {}
    
    func parse(tokens:[Token], symbols: SymbolsTable, debug: Bool) -> Bool {
        self.tokens = tokens
        self.symbols = symbols
        self.debug = debug
        
        return true
    }
    
    func match(name:String) -> Bool {
        if tokenIndex >= tokens.count {
            return false
        }
        
        let t: Token = tokens[tokenIndex]
        
        if t.type.name == name {
            tokenIndex += 1
            
            if debug {
                print("Match: \(name)")
            }
            
            return true
        }
        
        return false
    }
    
    func resetIndexes(tokenIndex:Int, tupleIndex:Int) -> Int {
        self.tokenIndex = tokenIndex
        deleteTuples(tupleIndex, end:tuples.count)
        return tuples.count
    }
    
    func deleteTuples(begin:Int, end:Int) {
        let n:Int = end - begin
        
        for _ in 0..<n {
            tuples.removeAtIndex(tuples.count - 1)
        }
    }
    
    func buildTuple(type:String, begin:Int, end:Int, nextTrue:Int, nextFalse:Int) {
        var elements: [Token] = []
        
        for i in begin..<end {
            elements.append(tokens[i])
        }
        
        tuples.append(Tuple(type:type, elements:elements, nextTrue:nextTrue, nextFalse:nextFalse))
    }
}